### ISSUE TITLE
Issues with ListBox, ListView, multiselect and SelectedValue

### ISSUE THREAD

**Issue by BentTranberg** [elmish/Elmish.WPF#440]
**edit: SPOILER ALERT! It seems I was wrong about SelectedValue not working sensibly. See further down below for a working demo. edit2: And I also got ListView multiselect to work. See even further down for a demo of that.**

This is not about problems with Elmish.WPF, but it is about experiences I've had that I believe will be highly relevant to users of Elmish.WPF that want to use SelectedValue in WPF, or multiselect, or a combination of these. I am absolutely amazed that these issues exists at all in WPF, and seems to have been there for years.

## SelectedValue not working sensibly

I had trouble with SelectedValue because the binding did not trigger when I unselected the last selected value in a multiselect (Extended kind) ListBox. (There are SO issues about this, and it seems hard to get around this.) And so the option value for SelectedValue in my model is not set to None, as it should have been.

But I found a workaround.

Since I am using multiselect in combination with SelectedValue, I found a workaround by using the SetIsSelected multiselect handler, by detecting whether this isSelected is the only remaining selected item to become unselected, and then setting the option value for SelectedValue to None at that point.

This problem is so nasty that I'm thinking that perhaps there should be a sample project demoing it.

## Trouble with multiselect and ListView

I also tried to use ListView with multiselect before giving up and resorting to ListBox. The reason for that trouble was that the IsSelected style setter property fails to work with ListView. (There are SO issues about this, and it seems hard to get it to work.) I didn't really need ListView in this case, but I hope to find a solution to this problem before I do have a case where the columns of ListView will be useful.

**Comment by TysonMN** [comment]
I am interested in adding such workarounds involving Elmish.WPF to the samples.

Are you able to share a link to a GitHub repository that reproduces this behavior?

**Comment by BentTranberg** [comment]
I want to extract a repro from the project at work where I've made this workaround for the first case. I've been overwhelmed with work so far this month, but hope to get a chance in the coming weekend.

**Comment by BentTranberg** [comment]
Turns out I was wrong about SelectedValue not working sensibly. I have no idea what I did wrong earlier, or whether some change has been done elsewhere since then - perhaps both. I've updated from Elmish.WPF 4.0.0-beta-41 to 42, the latest. This time there was an ERR in the log, and I don't remember seeing that earlier, but anyway it's far more likely it was all my fault.

Here is the minimal demo source I have for this now, which, just to be clear, now works as expected. It's a small zip with a self-contained solution (pulling Elmish.WPF and Serilog from NuGet though) with a single project, and source only.
[Elmish2022.zip](https://github.com/elmish/Elmish.WPF/files/7574762/Elmish2022.zip)

And now for the second issue. I am now going to look into use of ListView in place of ListBox, and otherwise identical to this demo. I think that issue is going to be harder for me to resolve, because there really was a lot of Qs on SO that didn't have ideal solutions, at least none that I was able to get working when I tried earlier.

**Comment by TysonMN** [comment]
> I've updated from Elmish.WPF 4.0.0-beta-41 to 42, the latest. This time there was an ERR in the log, and I don't remember seeing that earlier, but anyway it's far more likely it was all my fault.

Maybe it was [this](https://github.com/elmish/Elmish.WPF/blob/master/RELEASE_NOTES.md?plain=1#L11) or [this](https://github.com/elmish/Elmish.WPF/blob/master/RELEASE_NOTES.md?plain=1#L16) improvement.
> - Now logging when WPF tries to get or set a binding and an exception is thrown
> - Now logging and returning false to WPF if selection in a `SubModelSelectedItem` binding fails

**Comment by TysonMN** [comment]
> I am now going to look into use of ListView in place of ListBox, and otherwise identical to this demo. I think that issue is going to be harder for me to resolve, because there really was a lot of Qs on SO that didn't have ideal solutions, at least none that I was able to get working when I tried earlier.

Good luck!

If you share a reproduction, then I will take a look.

**Comment by BentTranberg** [comment]
Yes! That first one I guess. Thanks! These changes are very helpful indeed.

| - Now logging when WPF tries to get or set a binding and an exception is thrown

But I tried to put a catch all around the setter when I tried earlier, and it didn't catch anything, and it didn't even go into the setter if breakpoints could be trusted. Right this second I'm thinking perhaps I returned a value in the getter that somehow silenced the setter. Well, It's good now, so I'll stop thinking about it.

| If you share a reproduction, then I will take a look.

Ok. Not sure I can get anywhere at all with this one though.

**Comment by TysonMN** [comment]
Another debugging technique to gain more information about exceptions is to subscribe to the [FirstChanceExceptions](https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.firstchanceexception?view=net-6.0) event.

**Comment by BentTranberg** [comment]
That was quick! All I had to do was switch from `ListBox `to `ListView`, and use `ListView.View` instead of `ListBox.ItemTemplate`, and the minor changes that follows from that. Several issues I ran into earlier simply doesn't exist any more. This is amazing.

Now I'm getting seriously suspicious about what's going on. The Microsoft teams has been really hard at work lately fixing things all over the place, instead of just moving on with new stuff. Last I tried it was still previews of several things, but now it's all release from MS on my machine. But as already mentioned, I've also updated Elmish.WPF since last I tried. I am more inclined to believe you improved something there.

I notice that the demo of Multiselect has this

`<Style TargetType="{x:Type ListBoxItem}" BasedOn="{StaticResource {x:Type ListBoxItem}}">`

I threw out BasedOn in my ListBox demo, because it didn't seem to have a purpose.

When switching to ListView, BasedOn doesn't work at all, whether the type given is ListBoxItem or ListViewItem. So I _must_ leave it out, or else the items become this #%!ViewModel#%! (weird) stuff. The TargetType can be ListBoxItem also, and actually one of ListBoxItem or ListViewItem didn't work previously there but both does now. So...

`<Style TargetType="{x:Type ListViewItem}">`

Updated solution here, ListBox as earlier, ListView is new:
[Elmish2022-b.zip](https://github.com/elmish/Elmish.WPF/files/7574917/Elmish2022-b.zip)

I guess we close this now?
